### PR TITLE
Add a self-hosted shared cache example

### PR DESCRIPTION
**Pull request by better-salmon**
### What?

This pull request integrates the exemplary setup for a self-hosted Next.js application utilizing Redis as a shared cache storage. The solution supports caching at both the App and Pages routers in default and standalone modes, as well as partial pre-rendering, facilitated by the [`@neshca/cache-handler`](https://github.com/caching-tools/next-shared-cache/tree/canary/packages/cache-handler) package. The package enables customizing cache handlers and replacing the default cache provided by Next.js seamlessly.

### Why?

The motivation behind this pull request is to provide an example demonstrating how Redis can be used as a shared cache in a self-hosted environment, thereby improving the scalability of hosting multiple instances of a Next.js application.

### CODE REVIEW

**Comment by Fredkiss3** [comment]
You may want to wait for this #57953 before merging and update your PR when it has been merged

**Comment by better-salmon** [comment]
> You may want to wait for this #57953 before merging and update your PR when it has been merged

Hi @Fredkiss3, thank you for pointing out PR #57953. I had it in mind and should have mentioned it earlier. It's reasonable to wait for this feature, especially if it will be included in the upcoming 14.1.0 release by the Next.js team. I'll keep an eye on that PR, and once it's merged, I'll update my PR accordingly.

File: examples/self-hosted-shared-cache/README.md
Comment (leerob) [inline examples/self-hosted-shared-cache/README.md:3]: Could you explain why the custom package is needed versus just putting it into this example directly?

**Review by leerob — COMMENTED** [review]
Thank you for making this!

File: examples/self-hosted-shared-cache/README.md
Comment (better-salmon) [reply examples/self-hosted-shared-cache/README.md:3]: TLDR: this custom package takes on the responsibilities of the original class, providing a concise API for the user.

Since the API implies replacing the original FileSystemCache class with a custom one, it is not enough to replace the Map with the Redis Client from [the example in the docs](https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath). In this case, reading pre-rendered pages from disk will be lost. You can see [that my test is failing](https://github.com/caching-tools/next-shared-cache/actions/runs/6764512665/job/18382888425?pr=94#step:9:240) if I restrict file system access. The library implements work with the file system. So, when Redis fails, the applications in pods will continue to work with the file system until Redis is online.

Also, for the correct operation of on-demand revalidation, you must have a common TagsManifest for all pods. Since tagsManifest is checked in [this](https://github.com/vercel/next.js/blob/31c2f976cdd4deb0e7c412538a70b795dff4689e/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L255) and [this](https://github.com/vercel/next.js/blob/31c2f976cdd4deb0e7c412538a70b795dff4689e/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L275) places in the original class, it is necessary to support this functionality.

This functionality is the same for everyone and is, therefore, encapsulated in the library.

In an ideal world, I would like to have such an API for customizing caching out of the box, and I would be willing to spend the effort and time to implement this for Next.js since, in essence, my library is inspired mainly by your implementation.

Comment (better-salmon) [reply examples/self-hosted-shared-cache/README.md:3]: I will add this to example

Comment (better-salmon) [reply examples/self-hosted-shared-cache/README.md:3]: I also created this issue #58094. For now, it is crucial to read from disk using cache. This custom package deals with this out of the box.

If the disk is being used, key eviction alone cannot make on-demand revalidation; the users must also delete files. I hope this limitation is addressed soon, but now it works like so.

**Comment by leerob** [comment]
Went ahead and merged the improved self-hosting docs before this. Do you want to link back to the example in the updated docs in this PR?

**Comment by better-salmon** [comment]
> Went ahead and merged the improved self-hosting docs before this. Do you want to link back to the example in the updated docs in this PR?

@leerob, do you mean I can add a link or/and some description for this example to the updated docs by myself?

**Comment by better-salmon** [comment]
@leerob, please review the modifications I made in the documentation.

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 4e523bcf8557d7a66ce1fa4ea0193d49ad3d540a

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 4e523bcf8557d7a66ce1fa4ea0193d49ad3d540a

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Review by leerob — APPROVED** [review]
Nice work. Thank you for this contribution!

**Pull request merged**
